### PR TITLE
Update editorconfig Rules and Severity

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- Update rules in .editorconfig based on latest in dotnet/runtime
- Configure rules to trigger on build
- Combine rules from ruleset and and project files

## Related issues
[AB#81932](https://microsofthealth.visualstudio.com/Health/_workitems/edit/81932/)

## Testing
Pipelines pass and ad-hoc local builds to validate rules are catching problems.

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Api/Web/SeekableStreamConverter.cs [hunk 1]
```diff
@@ -56,31 +56,28 @@ public async Task<Stream> ConvertAsync(Stream stream, CancellationToken cancella
 
         private static class AspNetCoreTempDirectory
         {
-            private static string _tempDirectory;
+            private static string s_tempDirectory;
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/Web/SeekableStreamConverter.cs:59]: This is the default prefix for static members. How do people feel about `"s_"` prefixes?

File: src/Microsoft.Health.Dicom.Client/DicomWebClient.cs [hunk 2]
```diff
@@ -303,11 +303,15 @@ public async Task<DicomWebResponse> DeleteAsync(Uri requestUri, CancellationToke
             return new DicomWebResponse(response);
         }
 
-        public async Task<DicomWebAsyncEnumerableResponse<DicomDataset>> QueryAsync(string requestUri, CancellationToken cancellationToken)
+        [Obsolete("Please use QueryAsync(Uri, CancellationToken) instead.")]
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Client/DicomWebClient.cs:306]: We can change this method to use `Uri` in a backwards compatible manner. But we can also just delete the `string` overload.

File: src/Microsoft.Health.Dicom.Client/IDicomWebClient.cs [hunk 3]
```diff
@@ -24,8 +24,11 @@ public interface IDicomWebClient
 
         Task<DicomWebResponse<ChangeFeedEntry>> GetChangeFeedLatest(string queryString = "", CancellationToken cancellationToken = default);
 
+        [Obsolete("Please use QueryAsync(Uri, CancellationToken) instead.")]
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Client/IDicomWebClient.cs:27]: Can just delete this overload too.

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/Resources/Retrieve/LazyTransformReadOnlyStreamTests.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@ namespace Microsoft.Health.Dicom.Core.UnitTests.Features.Resources.Retrieve
 {
     public class LazyTransformReadOnlyStreamTests
     {
-        private static readonly RecyclableMemoryStreamManager _recyclableMemoryStreamManager = new RecyclableMemoryStreamManager();
+        private static readonly RecyclableMemoryStreamManager RecyclableMemoryStreamManager = new RecyclableMemoryStreamManager();
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core.UnitTests/Features/Resources/Retrieve/LazyTransformReadOnlyStreamTests.cs:17]: Do we want `static readonly` fields to use the same rules:
- `const` (PascalCase)
- `private` (`"_example"`)
- `static private` (`"s_example"`)

File: src/Microsoft.Health.Dicom.Core/Features/Context/DicomRequestContext.cs [hunk 5]
```diff
@@ -13,36 +14,41 @@ namespace Microsoft.Health.Dicom.Core.Features.Context
...
     public class DicomRequestContext : IDicomRequestContext
     {
-        private readonly string _uriString;
-        private readonly string _baseUriString;
-
-        private Uri _uri;
-        private Uri _baseUri;
-
+        [Obsolete("Please use the other constructor.")]
+        [SuppressMessage("Design", "CA1054:URI-like parameters should not be strings", Justification = "Legacy constructor.")]
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Context/DicomRequestContext.cs:18]: This is probably the biggest change in here.

File: test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/TestDicomWebServerFactory.cs [hunk 6]
```diff
@@ -19,12 +18,12 @@ public static TestDicomWebServer GetTestDicomWebServer(Type startupType)
                 return new InProcTestDicomWebServer(startupType);
             }
 
-            if (environmentUrl.Last() != '/')
+            if (environmentUrl[^1] != '/')
```
Comment (wsugarman) [inline test/Microsoft.Health.Dicom.Web.Tests.E2E/Common/TestDicomWebServerFactory.cs:21]: New syntax!

File: .editorconfig [hunk 7]
```diff
@@ -10,102 +10,122 @@ root = true
...
+# Modifier preferences
+dotnet_style_require_accessibility_modifiers = for_non_interface_members:error
+csharp_preferred_modifier_order              = public,private,protected,internal,static,extern,new,virtual,abstract,sealed,override,readonly,unsafe,volatile,async:suggestion
 
 # avoid this. unless absolutely necessary
-dotnet_style_qualification_for_field = false:suggestion
-dotnet_style_qualification_for_property = false:suggestion
-dotnet_style_qualification_for_method = false:suggestion
-dotnet_style_qualification_for_event = false:suggestion
+dotnet_style_qualification_for_field    = false:error
```
Comment (wsugarman) [inline .editorconfig:46]: These still don't seem to trigger a build error, so I'll file a bug about that with the roslyn team.

File: .editorconfig [hunk 8]
```diff
@@ -124,71 +146,97 @@ csharp_style_inlined_variable_declaration = true:suggestion
...
 
-# Wrapping Options
-csharp_preserve_single_line_blocks = true
-csharp_preserve_single_line_statements = false
-
-# C++ Files
-
-# IDE0052: Remove unread private members
-dotnet_diagnostic.IDE0052.severity = warning
+# Analyzers
```
Comment (wsugarman) [inline .editorconfig:178]: Everything under "Analyzers" includes the rules previously imposed by the ruleset files

File: src/Microsoft.Health.Dicom.Api/Web/SeekableStreamConverter.cs (hunk 1 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Api/Web/SeekableStreamConverter.cs:59]: Based on my discussions with folks, we're ok with this!

File: src/Microsoft.Health.Dicom.Core.UnitTests/Features/Resources/Retrieve/LazyTransformReadOnlyStreamTests.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core.UnitTests/Features/Resources/Retrieve/LazyTransformReadOnlyStreamTests.cs:17]: Combining `const` with `static` `readonly` styles

File: .editorconfig [hunk 9]
```diff
@@ -10,110 +10,140 @@ root = true
 insert_final_newline = true
```
Comment (wsugarman) [inline .editorconfig:10]: Mostly important are the changes here

File: .editorconfig (hunk 7 above)
Comment (wsugarman) [reply .editorconfig:46]: Issue is filed here: https://github.com/dotnet/roslyn-analyzers/issues/5088

Comment (wsugarman) [reply .editorconfig:46]: I included the details from this issue in the file itself

File: .editorconfig [hunk 10]
```diff
@@ -10,110 +10,140 @@ root = true
...
 indent_style = space
 indent_size = 4
+trim_trailing_whitespace = true
 
 [project.json]
 indent_size = 2
 
+# Generated code
+[*{_AssemblyInfo.cs,.notsupported.cs}]
```
Comment (brandonpollett) [inline .editorconfig:19]: Do we need to include `.Generated.cs`? For example: https://github.com/microsoft/dicom-server/blob/main/src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Model/V1.Generated.cs

File: .editorconfig [hunk 11]
```diff
@@ -124,71 +154,104 @@ csharp_style_inlined_variable_declaration = true:suggestion
...
 # Shell scripts
 [*.sh]
 end_of_line = lf
-[*.{cmd, bat}]
+[*.{cmd,bat}]
 end_of_line = crlf
 
 # YAML files
-[*.{yml, yaml}]
+[*.{yml,yaml}]
```
Comment (brandonpollett) [inline .editorconfig:256]: This seems to be in here twice.

File: src/Microsoft.Health.Dicom.Client/DicomWebClient.cs (hunk 2 above)
Comment (brandonpollett) [reply src/Microsoft.Health.Dicom.Client/DicomWebClient.cs:306]: I think we can leave it for now. I do wonder if we should make the documentation that says it needs to be relative.

**Review by brandonpollett — APPROVED** [review]

File: .editorconfig (hunk 10 above)
Comment (wsugarman) [reply .editorconfig:19]: Good catch

**Pull request merged**